cranelift/codegen/src/loop_analysis.rs

```diff
@@ -228,6 +283,28 @@ impl LoopAnalysis {
             }
         }
     }
+
+    fn assign_loop_levels(&mut self) {
+        let mut stack: SmallVec<[Loop; 8]> = smallvec![];
+        for lp in self.loops.keys() {
+            if self.loops[lp].level == LoopLevel::invalid() {
+                stack.push(lp);
+                while let Some(&lp) = stack.last() {
+                    if let Some(parent) = self.loops[lp].parent.into() {
+                        if self.loops[parent].level != LoopLevel::invalid() {
+                            self.loops[lp].level = self.loops[parent].level.inc();
+                            stack.pop();
+                        } else {
+                            stack.push(parent);
+                        }
+                    } else {
+                        self.loops[lp].level = LoopLevel::root().inc();
+                        stack.pop();
+                    }
+                }
+            }
+        }
```

Comment:
I think the following implementation is correct. Asymptotically, it has the same time complexity, but its space complexity is O(1) instead of O(n). Whether it's faster or slower in practice is another question entirely, of course. My guess is it's about the same.

```suggestion
        for lp in self.loops.keys() {
            if self.loops[lp].level == LoopLevel::invalid() {
                let mut assigned = self.loops[lp].parent;
                let mut level = 0usize;
                while let Some(parent) = assigned.expand() {
                    level += 1;
                    if self.loops[parent].level != LoopLevel::invalid() {
                        level += self.loops[parent].level.0.into();
                        break;
                    }
                    assigned = self.loops[parent].parent;
                }
                let mut cur = PackedOption::from(lp);
                while cur != assigned {
                    self.loops[cur.unwrap()].level = level.into();
                    cur = self.loops[cur.unwrap()].parent;
                    level -= 1;
                }
            }
        }
```

In writing this, though, I realized that `LoopLevel::root()` can't appear in `loops[lp].level` after analysis is done, which means I think zero would be a better choice for an "invalid" value. This would all be a little simpler if a loop level were represented as `Option<NonZeroU8>`. Then you can even replace `LoopLevel::inc` with `NonZeroU8::saturating_add`, which apparently was just added in Rust 1.64.

Reply:
Ah, I see, this is the same trick we saw in the famous single-predecessor-chain traversals :-) OK, yeah, avoiding the `SmallVec` might be worthwhile here. I agree that the shallow loop nests are unlikely to stress either algorithm (so e.g. pointer-chasing overhead of traversing twice is likely negligible) so on balance let's go with yours as it requires less storage.

Reply:
Yup, same trick! I might be over-eager to apply that trick now, but it's a pretty good trick. 😁

I hope my version works; I meant it to be illustrative, more than an exact solution. I typed it directly into the GitHub comment, so I didn't even know whether it would compile…

I was briefly tempted to suggest a different trick: temporarily overwrite the parent pointers during the traversal so they instead point to the child you need to return to, then put back the original parent pointers as you walk back up the stack. That gives you the stack traversal pattern in O(1) additional space. It's also too clever/confusing to justify using here, I think. 😅 I believe I first learned about it in the context of some garbage collector, where it makes more sense.

Reply:
It almost worked! The last `level -= 1` would underflow before we take the backedge and hit a loop-exiting condition; reconfigured things a bit to break before that.

Reply:
I think that means I got the edge cases wrong a different way: the smallest level it should assign is 1, so decrementing it an extra time in the last loop iteration should never underflow.

Also, I see you changed my usize to a u8 so you could directly construct a LoopLevel. I'd imagined a From impl that would saturate when converting from a usize that was too large. But this is fine too, at least in debug builds where overflow panics. Loop depth comes from untrusted input so I think that probably should be checked in release builds too, right? Restricting to 255-deep nested loops seems reasonable to me, anyway.